plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableMetaQuery.php

```diff
@@ -0,0 +1,566 @@
+<?php
+namespace Automattic\WooCommerce\Internal\DataStores\Orders;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class used to implement meta queries for the orders table datastore via {@see OrdersTableQuery}.
+ * Heavily inspired by WordPress' own `WP_Meta_Query` for backwards compatibility reasons.
+ *
+ * Parts of the implementation have been adapted from {@link https://core.trac.wordpress.org/browser/tags/6.0.1/src/wp-includes/class-wp-meta-query.php}.
+ */
+class OrdersTableMetaQuery {
+
+	/**
+	 * List of non-numeric SQL operators used for comparisons in meta queries.
+	 *
+	 * @var array
+	 */
+	private const NON_NUMERIC_OPERATORS = array(
+		'=',
+		'!=',
+		'LIKE',
+		'NOT LIKE',
+		'IN',
+		'NOT IN',
+		'EXISTS',
+		'NOT EXISTS',
+		'RLIKE',
+		'REGEXP',
+		'NOT REGEXP',
+	);
+
+	/**
+	 * List of numeric SQL operators used for comparisons in meta queries.
+	 *
+	 * @var array
+	 */
+	private const NUMERIC_OPERATORS = array(
+		'>',
+		'>=',
+		'<',
+		'<=',
+		'BETWEEN',
+		'NOT BETWEEN',
+
+	);
+
+	/**
+	 * Prefix used when generating aliases for the metadata table.
+	 *
+	 * @var string
+	 */
+	private const ALIAS_PREFIX = 'meta';
+
+	/**
+	 * Sanitized `meta_query`.
+	 *
+	 * @var array
+	 */
+	private $queries = array();
+
+	/**
+	 * JOIN clauses to add to the main SQL query.
+	 *
+	 * @var array
+	 */
+	private $join = array();
+
+	/**
+	 * WHERE clauses to add to the main SQL query.
+	 *
+	 * @var array
+	 */
+	private $where = array();
+
+	/**
+	 * Table aliases in use by the meta query. Used to optimize JOINs when possible.
+	 *
+	 * @var array
+	 */
+	private $table_aliases = array();
+
+	/**
+	 * The meta_query relation.
+	 *
+	 * @var string
+	 */
+	private $relation = 'AND';
+
+	/**
+	 * Constructor.
+	 *
+	 * @param OrdersTableQuery $q The main query being performed.
+	 */
+	public function __construct( OrdersTableQuery $q ) {
+		$meta_query = $q->get( 'meta_query' );
+
+		if ( ! $meta_query ) {
+			return;
+		}
+
+		$this->queries  = $this->sanitize_meta_query( $meta_query );
+		$this->relation = $this->sanitize_relation( $meta_query['relation'] ?? 'AND' );
+
+		$this->meta_table   = $q->get_table_name( 'meta' );
+		$this->orders_table = $q->get_table_name( 'orders' );
+
+		$this->build_query();
+	}
+
+	/**
+	 * Returns JOIN and WHERE clauses to be appended to the main SQL query.
+	 *
+	 * @return array {
+	 *     @type string $join  JOIN clause.
+	 *     @type string $where WHERE clause.
+	 * }
+	 */
+	public function get_sql_clauses() {
+		return array(
+			'join'  => $this->sanitize_join( $this->join ),
+			'where' => $this->flatten_where_clauses( $this->where ),
+		);
+	}
+
+	/**
+	 * Checks whether a given meta_query clause is atomic or not (i.e. not nested).
+	 *
+	 * @param array $arg The meta_query clause.
+	 * @return boolean TRUE if atomic, FALSE otherwise.
+	 */
+	private static function is_atomic( $arg ) {
```

Comment:
why is this function `static` when no other method in the class is.

Reply:
No particular reason. It's just because its result didn't depend at all on the internal state, but it'd be best to make it non-static as it serves no real purpose that way.